packages/flutter/lib/src/widgets/image_filter.dart

```diff
@@ -46,32 +65,75 @@ class ImageFiltered extends SingleChildRenderObjectWidget {
   }
 }
 
-class _ImageFilterRenderObject extends RenderProxyBox {
-  _ImageFilterRenderObject(this._imageFilter);
+/// Applies an [ImageFilter] to its child.
+class RenderImageFiltered extends RenderProxyBox {
+  /// Creates a [RenderObject] that applies an [ImageFilter] to its child.
+  ///
+  /// One of the [imageFilter] or [imageFilterCallback] should not be null here
+  /// or whenever the [paint] method is invoked.
+  /// The [imageFilter] will be used if it is not null, otherwise the filter
+  /// will be obtained from the [imageFilterCallback].
+  RenderImageFiltered({
+    required ImageFilter? imageFilter,
+    required ImageFilterCallback? imageFilterCallback,
+  })
+      : assert(imageFilter != null || imageFilterCallback != null),
+        _imageFilter = imageFilter,
+        _imageFilterCallback = imageFilterCallback;
 
-  ImageFilter get imageFilter => _imageFilter;
-  ImageFilter _imageFilter;
-  set imageFilter(ImageFilter value) {
-    assert(value != null);
+  /// The [ImageFilter] to apply to this child, or null if the filter will be supplied
+  /// instead by the [imageFilterCallback].
+  ///
+  /// If the [imageFilter] is set to null here, then either it or the [imageFilterCallback]
+  /// should be set to a non-null value before [paint] is called.
```

Comment:
docs should say what happens when both are non-null too

i'd probably just allow both to be null and both to be non-null and define which wins if they're both non-null and say it's a no-op if they're both null, since you need to be able to handle the setters being called in either order and thus can't enforce the widget's contract here.

Reply:
That's an interesting alternative. I'll think about it.

One thing that bothers me about these NOP conditions is that they can lead to lower level layers coming and going that affect output. We already have an issue filed on the fact that omitting the opacity layer when the opacity is 1.0 can change output, sometimes drastically.

Reply:
I decided that the conditions for the widget and the render object should be consistent, so the render object asserts that only one is specified in both constructor and paint.